cpp/src/parquet/encoding.cc

```diff
@@ -442,7 +442,13 @@ class DictEncoderImpl : public EncoderImpl, virtual public DictEncoder<DType> {
         dict_encoded_size_(0),
         memo_table_(pool, kInitialHashTableSize) {}
 
-  ~DictEncoderImpl() override { DCHECK(buffered_indices_.empty()); }
+  ~DictEncoderImpl() override {
+#ifndef NDEBUG
+    if (!buffered_indices_.empty()) {
+      ARROW_LOG(WARNING) << "DictEncoderImpl destroyed without flushing or closing";
```

Comment:
```suggestion
      ARROW_LOG(WARNING) << "DictEncoderImpl is destroyed without flushing or closing";
```

Reply:
IIUC, the behavior has changed under debug mode, which calls abort() in the past.

Reply:
The reason is described here: https://github.com/apache/arrow/issues/38117